compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -208,7 +208,18 @@ class ClassCodegen private constructor(
             entry is MultifileFacadeFileEntry -> KotlinClassHeader.Kind.MULTIFILE_CLASS
             else -> KotlinClassHeader.Kind.SYNTHETIC_CLASS
         }
-        writeKotlinMetadata(visitor, state, kind, extraFlags) {
+
+        // There are three kinds of classes which are regenerated during inlining.
+        // 1) Anonymous classes which are in the scope of an inline function.
+        // 2) SAM wrappers used in an inline function. These are identified by name, since they
+        //    can be reused in different functions and are thus generated in the enclosing top-level
+        //    class instead of inside of an inline function.
+        // 3) WhenMapping classes used from inline functions. These are collected in
+        //    `JvmBackendContext.publicAbiSymbols` in `MappedEnumWhenLowering`.
```

Comment:
In future we can support lambda/sam/function reference generation via invokedynamic. Would it affect current scheme in any sense? Should we write any flags for bootstrap methods?

Reply:
The bootstrap method should be unaffected, since it probably won't contain any anonymous objects. The only potential problem that I can think of is that the invoke method will be lowered into a private method in the enclosing class. We should ensure that we add a reference to the original `IrFunction` via `attributeOwnerId` so that `isInPublicInlineScope` keeps working.